subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformerInvocationFactory.java

```diff
@@ -125,9 +125,9 @@ public DefaultTransformerInvocationFactory(
         FileCollectionFingerprinter outputFingerprinter = fingerprinterRegistry.getFingerprinter(OutputNormalizer.class);
         FileCollectionFingerprinter dependencyFingerprinter = fingerprinterRegistry.getFingerprinter(transformer.getInputArtifactDependenciesNormalizer());
 
-        MerkleDirectorySnapshotBuilder builder = MerkleDirectorySnapshotBuilder.sortingRequired();
-        virtualFileSystem.read(inputArtifact.getAbsolutePath(), builder);
-        FileSystemLocationSnapshot inputArtifactSnapshot = builder.getResult();
+        MutableReference<FileSystemLocationSnapshot> inputArtifactSnapshotReference = MutableReference.empty();
+        virtualFileSystem.read(inputArtifact.getAbsolutePath(), inputArtifactSnapshotReference::set);
+        FileSystemLocationSnapshot inputArtifactSnapshot = inputArtifactSnapshotReference.get();
```

Comment:
We should have a version of `read()` take a `Function` and return a `<T>` instead of doing this everywhere.

Reply:
I added a method `read(location, function)` and removed the `read(location, consumer)` method. When reading the location without a filter, we always want the result. When reading a location with the filter, we consume the result if available.

Reply:
I think this is okay for now. I'd like to end up with a more consolidated API eventually.